File: python/linked-list/remove_elements.py
# Given the head of a linked list and an integer val, remove all the nodes of the linked list that has Node.val == val, and return the new head.

# Implementing a List Node
class Node:
    def __init__(self, val) -> None:
        self.val = val
        self.next = None

# Implementing a Linked List
class LinkedList:
    def __init__(self, head) -> None:
        self.head = head
        self.tail = head
    
    def add(self, node):
        if not self.head:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node

node1 = Node(1)
node2 = Node(2)
node3 = Node(6)
node4 = Node(3)
node5 = Node(4)
node6 = Node(5)
node7 = Node(6)

linkedList = LinkedList(node1)
linkedList.add(node2)
linkedList.add(node3)
linkedList.add(node4)
linkedList.add(node5)
linkedList.add(node6)
linkedList.add(node7)

def removeElements(head: Node, val: int) -> Node:
    dummy = Node()
    dummy.next = head
    prev = dummy
    curr = head

    while curr:
        if curr.val == val:
            prev.next = curr.next
            curr = curr.next
        else:
            prev = curr
            curr = curr.next
    
    return dummy.next

# tested on leetcode.com/problems/remove-linked-list-element

File: javascript/linked-list/remove-elements.js
// Remove Linked List Elements

// Given the head of a linked list and an integer val, remove all the nodes of the linked list that has Node.val == val, and return the new head.

// impelemnting node 
class ListNode {
    constructor(val) {
        this.val = val
        this.next = null
    }
}

// Implementing a Linked List
class LinkedList {
    constructor(head = null) {
        this.head = head
        this.tail = head
    }

    add(node) {
        if (!this.head) {
            this.head = node
            this.tail = node
        } else {
            this.tail.next = node
            this.tail = node
        }
    }
}
// initialize the nodes
let node1 = new ListNode(1)
let node2 = new ListNode(2)
let node3 = new ListNode(6)
let node4 = new ListNode(3)
let node5 = new ListNode(4)
let node6 = new ListNode(5)
let node7 = new ListNode(6)
// initilize the linked lists
const list = new LinkedList(node1)
list.add(node2)
list.add(node3)
list.add(node4)
list.add(node5)
list.add(node6)
list.add(node7)

function removeElements(head, val) {
    let dummy = new ListNode()
    dummy.next = head
    let prev = dummy
    let curr = head

    while (curr) {
        if (curr.val === val) {
            prev.next = curr.next
            curr = curr.next
        } else {
            prev = curr
            curr = curr.next
        }
    }

    return dummy.next
}

// testing
console.log(removeElements(list.head, 6))
// expected output:
// ListNode(1) -> ListNode(2) -> ListNode(3) -> ... 